File: app/src/main/java/com/example/challenge_consult_cep/feature/search_cep/data/remote/SearchCepDataSourceRemote.java
package com.example.challenge_consult_cep.feature.search_cep.data.remote;

import com.example.challenge_consult_cep.feature.search_cep.data.mapper.ApiCepDTOMapper;
import com.example.challenge_consult_cep.feature.search_cep.data.mapper.AwesomeApiDTOMapper;
import com.example.challenge_consult_cep.feature.search_cep.data.mapper.ViaCepDTOMapper;
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.ApiCepApi;
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.AwesomeApiApi;
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.ViaCepApi;
import com.example.challenge_consult_cep.feature.search_cep.domain.model.Cep;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

public class SearchCepDataSourceRemote {

    private ApiCepApi apiCepApi;
    private AwesomeApiApi awesomeApiApi;
    private ViaCepApi viaCepApi;

    public SearchCepDataSourceRemote(ApiCepApi apiCepApi, AwesomeApiApi awesomeApiApi, ViaCepApi viaCepApi) {
        this.apiCepApi = apiCepApi;
        this.awesomeApiApi = awesomeApiApi;
        this.viaCepApi = viaCepApi;
    }

    public Single<Cep> getCep(String cep) {

        String cepWithHyphen = cep.substring(0, 5) + "-" + cep.substring(5, 8);

        return viaCepApi.getCep(cep).map(ViaCepDTOMapper::toCep)
                .onErrorResumeNext(one -> apiCepApi.getCep(cepWithHyphen).map(ApiCepDTOMapper::toCep))
                .onErrorResumeNext(two -> awesomeApiApi.getCep(cep).map(AwesomeApiDTOMapper::toCep));
    }
}

File: app/src/main/java/com/example/challenge_consult_cep/feature/search_cep/data/remote/SearchCepDataSourceRemote.kt
package com.example.challenge_consult_cep.feature.search_cep.data.remote

import com.example.challenge_consult_cep.feature.search_cep.data.mapper.toCep
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.ApiCepApi
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.AwesomeApiApi
import com.example.challenge_consult_cep.feature.search_cep.data.remote.api.ViaCepApi
import com.example.challenge_consult_cep.feature.search_cep.domain.model.Cep

class SearchCepDataSourceRemote(
    private val apiCepApi: ApiCepApi,
    private val awesomeApiApi: AwesomeApiApi,
    private val viaCepApi: ViaCepApi,
) {

    suspend fun getCep(cep: String): Cep {
        return try {
            return viaCepApi.getCep(cep).toCep()
        } catch (e: Exception) {
            try {
                val newCep = cep.substring(0, 5) + "-" + cep.substring(5, 8)
                return apiCepApi.getCep(newCep).toCep()
            } catch (e: Exception) {
                try {
                    return awesomeApiApi.getCep(cep).toCep()
                } catch (e: Exception) {
                    throw Exception(e)
                }
            }
        }
    }
}